File: 0120-triangle/0120-triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int m = triangle.size();

        vector<vector<int>> dp(m, vector<int>(m, -1));
        
        for(int j = 0; j < m; j++) {
            dp[m - 1][j] = triangle[m - 1][j];
        }
        
        for(int j = m - 2; j >= 0; j--) {
            for(int i = 0; i <= j; i++) {
                int sameColumnNextRow = triangle[j][i] + dp[j + 1][i];
                int nextColumnNextRow = triangle[j][i] + dp[j + 1][i + 1];

                dp[j][i] = min(sameColumnNextRow, nextColumnNextRow);
            }
        }
        
        return dp[0][0];
    }
};

File: 0120-triangle/0120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        int[] dp = new int[n];

        // Initialize dp with the last row of the triangle
        for (int j = 0; j < triangle.get(n - 1).size(); j++) {
            dp[j] = triangle.get(n - 1).get(j);
        }

        // Bottom-up DP computation
        for (int i = n - 2; i >= 0; i--) {
            int[] cur = new int[n];
            for (int j = i; j >= 0; j--) {
                cur[j] = triangle.get(i).get(j) + Math.min(dp[j], dp[j + 1]);
            }
            dp = cur;
        }
        
        return dp[0];
    }
}